[ENT-492] Add initial implementation of a generic Consent API.

UmanShahzad opened the pull request:
**Description:** (to quote the upstream ticket): Right now, our consent architecture is fragmented across the `EnterpriseCustomerUser` API and the `EnterpriseCourseEnrollment` API. This PR creates an API that can handle multiple use cases, based on the existing set of data models, in a new consent Djangoapp.

**JIRA:** [ENT-492](https://openedx.atlassian.net/browse/ENT-492)

**Dependencies:** None, but future stories depend on this.

**Merge deadline:** End of this week.

**Testing instructions:**

1. Confirm that Consent API tests cover all important cases.
1. Install this version of `edx-enterprise` in `edx-platform` on a devstack.
1. Register an account on that devstack.
1. Make an `EnterpriseCustomer`.
1. Make an `EnterpriseCustomerUser` based off of your registered account and associate it with the customer from the previous step.
1. Configure a course for the `EnterpriseCustomer` with whatever consent requirements you plan on testing.
1. Enroll in the course, and either provide or do not provide consent for that course, depending on what you plan on testing.
1. Send queries to endpoint `/consent/api/v1/data_sharing_consent` in accordance with the `username` of the account you created, the UUID of the `enterprise_customer` you created, and the `course_id` of the course you enrolled in. Ensure that consent-related responses match with what you planned on testing.
1. If you did not provide consent, send a `POST` request using the `curl` command line tool to the same endpoint (`/consent/api/v1/data_sharing_consent`) and confirm that consent is granted for the course.
1. If you did provide consent, send a `DELETE` request using the `curl` command line tool to the same endpoint (`/consent/api/v1/data_sharing_consent`) and confirm that consent is revoked for that course.

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [x] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

**Author concerns:** 

1. This is a stepping stone to something more complete; currently we're using models from the `enterprise` app, and we'd like to eventually use models internal to the `consent` app.
1. **The first commit of this application reorganizes the tests to make it easier and more sensible to add tests for new applications.**

**Reviewers**
- @haikuginger 
- @brittneyexline @douglashall

edx-webhook commented:
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=151"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

haikuginger commented on tests/consent/api/test_views.py:
```diff
@@ -0,0 +1,452 @@
+        ),
+    )
+    @ddt.unpack
+    def test_consent_api_get_endpoint(self, factory, items, request_body, expected_response_body, expected_status_code):
```
Let's add checks for the different cases of EnterpriseCustomer consent requirement as well.

UmanShahzad replied:
I will do that in a coming push.

haikuginger commented on tests/consent/api/test_views.py:
```diff
@@ -0,0 +1,452 @@
+            {
+                'username': TEST_USERNAME,
+                'enterprise_customer': TEST_UUID,
+                'course_id': TEST_COURSE,
```
Why aren't we including state, etc, in our expectations for this test?

UmanShahzad replied:
So this is basically because I haven't actually made `POST` return any useful information. This is something I probably should have asked/discussed earlier: should `POST` return the following 2 extra keys:

```json
{
    "created": boolean,
    "consent_provided": boolean,
}
```

In my opinion, if `POST` returns a 200 status code, then it's obvious that `'consent_provided'` is going to be `True`, because we were able to successfully perform the only function relevant to `POST` here (provide consent), or create the resource and then perform that function. And if it returns 404, then `False`, because the request format was incorrect and we couldn't identify nor create the resource and provide consent. So do we need to return it? Either way, what do you reckon we should be returning to indicate state besides these?

haikuginger replied:
Let's return a value indicating the state as we would for a GET request with the given parameters.

UmanShahzad replied:
Alrighty!

haikuginger commented on tests/consent/api/test_views.py:
```diff
@@ -0,0 +1,452 @@
+            {
+                'username': TEST_USERNAME,
+                'enterprise_customer': TEST_UUID,
+                'course_id': TEST_COURSE,
```
As above. We need actual expectations of what the end status ought to be.

UmanShahzad replied:
See above comment for discussion / question.

haikuginger commented on tests/consent/test_helpers.py:
```diff
@@ -0,0 +1,41 @@
+    """
+
+    def test_consent_exists(self):
+        pass
```
I'm assuming these will be filled in. 😉

UmanShahzad replied:
Of course! Next on my agenda.

haikuginger commented on tox.ini:
```diff
@@ -23,17 +23,22 @@ match-dir = (?!migrations)
 commands =
+    touch tests/__init__.py
     py.test -Wd {posargs}
+    rm tests/__init__.py
```
Why?

UmanShahzad replied:
With the changes to the testing organization, I need to temporarily be able to access the `tests` package in some imports. Note that the same thing is being done below for quality testing. I'm not 100% sure why we don't just keep the `__init__.py`, though.

UmanShahzad replied:
Hmm, OK so I read this carefully:

```
setup.py
mypkg/
    ...
tests/
    __init__.py
    foo/
        __init__.py
        test_view.py
    bar/
        __init__.py
        test_view.py

Now pytest will load the modules as tests.foo.test_view and tests.bar.test_view, allowing you to have modules with the same name. But now this introduces a subtle problem: in order to load the test modules from the tests directory, pytest prepends the root of the repository to sys.path, which adds the side-effect that now mypkg is also importable. This is problematic if you are using a tool like tox to test your package in a virtual environment, because you want to test the installed version of your package, not the local code from the repository.

In this situation, it is strongly suggested to use a src layout where application root package resides in a sub-directory of your root:

setup.py
src/
    mypkg/
        __init__.py
        app.py
        view.py
tests/
    __init__.py
    foo/
        __init__.py
        test_view.py
    bar/
        __init__.py
        test_view.py

This layout prevents a lot of common pitfalls and has many benefits, which are better explained in this excellent blog post by Ionel Cristian Mărieș.
```

And for that reason, since I don't think I should be making a new `src/` directory (or some other appropriate name) as part of this PR, I'm moving `utils` back out to root as `test_utils` to avoid the `__init__.py`. In a future PR we can solve this problem.

bradenmacdonald commented on consent/__init__.py:
```diff
@@ -0,0 +1,3 @@
+"""
+edX Enterprise's Consent application.
```
In this, and perhaps also in some other docstrings for this app, can we please say more explicitly what "Consent" is? It's a very generic word and if I weren't part of the enterprise team I would have a hard time figuring out what this "consent" app is about.

Something like:
"The consent application provides an API that enables the platform to prompt the user to consent to share data about their activity and performance in (a) particular course(s) with the enterprise customer that is (usually) subsidizing their enrollment in those courses. When users grant consent to share data, that agreement is recorded in the database in an easily auditable way."

UmanShahzad replied:
Yeah the docstrings definitely need work, I just didn't want to spend too much time on them just yet, so thank you for this example; it looks good enough to be straight dropped right into it. I'll edit the rest to reflect something similar in a shorter fashion where appropriate.

UmanShahzad commented on consent/helpers.py:
```diff
@@ -0,0 +1,162 @@
+    return provided is not None and provided is not False
+
+
+def consent_required(user_id, course_id, enterprise_customer_uuid):
```
Yeah at first I had a pretty bad understanding of what the relation was between consent being required and consent being provided. Can you double check the changes I made to some of the existing test's `consent_required` expectations to see if they make sense?

UmanShahzad commented on consent/helpers.py:
```diff
@@ -0,0 +1,162 @@
+        user_data_consent_audit = get_user_dsc_audit(user_id, enterprise_customer)
+        provided = user_data_consent_audit is not None and user_data_consent_audit.enabled
+
+    return provided is not None and provided is not False
```
Yep, I had changed this function up before and forgot to change this.

UmanShahzad commented on consent/helpers.py:
```diff
@@ -0,0 +1,162 @@
+    enterprise_course_enrollment = get_enterprise_course_enrollment(user_id, course_id, enterprise_customer)
+    provided = enterprise_course_enrollment is not None and enterprise_course_enrollment.consent_available
+
+    if not provided or enterprise_course_enrollment is None:
```
Got it.

UmanShahzad commented on consent/helpers.py:
```diff
@@ -0,0 +1,162 @@
+    enterprise_customer = get_enterprise_customer(enterprise_customer_uuid)
+    enterprise_course_enrollment = get_enterprise_course_enrollment(user_id, course_id, enterprise_customer)
+
+    if enterprise_course_enrollment is None or not enterprise_course_enrollment.consent_available:
```
I'm quite happy with getting rid of the NoneType check :+1:

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,214 @@
+            raise ConsentAPIRequestError
+        enterprise_customer = get_enterprise_customer(enterprise_customer_uuid)
+        enterprise_customer_user = get_enterprise_customer_user(enterprise_customer, user_id)
+        enterprise_course_enrollment = get_enterprise_course_enrollment(user_id, course_id, enterprise_customer)
```
Did it, thanks!

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,214 @@
+
+    def get_required_query_params(self, request):
+        """
+
```
Yep, gotta fill this and some others.

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,214 @@
+            >>> }
+
+            The API accepts JSON objects with these key-value pairs for
+            POST or DELETE.
```
I don't think we should be doing that, see https://stackoverflow.com/questions/2539394/rest-http-delete-and-parameters for some reasons. I also tried it, and it broke a lot of tests, probably because this isn't really how it's meant to be used. I don't think we should have `DELETE` depend on query parameters.

brittneyexline replied:
I agree with Uman here - I don't typically see DELETE calls take query parameters.

brittneyexline commented on consent/helpers.py:
```diff
@@ -0,0 +1,149 @@
+    """
+    enterprise_customer = get_enterprise_customer(enterprise_customer_uuid)
+    enterprise_course_enrollment = get_enterprise_course_enrollment(user_id, course_id, enterprise_customer)
+    provided = enterprise_course_enrollment is not None and enterprise_course_enrollment.consent_available
```
nit: why not just `enterprise_course_enrollment and enterprise_course_enrollment.consent_available`?

UmanShahzad replied:
No special reason; I feel that reading it with "is not None" at the end there makes it more clear what I'm looking for.

UmanShahzad replied:
Scratch that: I just remembered we're trying to return a boolean here so it has to be `is not None` or else it'd return an object.

brittneyexline commented on consent/helpers.py:
```diff
@@ -0,0 +1,149 @@
+    enterprise_course_enrollment = get_enterprise_course_enrollment(user_id, course_id, enterprise_customer)
+    provided = enterprise_course_enrollment is not None and enterprise_course_enrollment.consent_available
+
+    if enterprise_course_enrollment is None:
```
IIRC it's also possible that `enterprise_course_enrollment.consent_available` can be None, and in that case it also means that we should look at the audit table.

UmanShahzad replied:
Ah, true! Yes I'll add to that.

brittneyexline commented on consent/helpers.py:
```diff
@@ -0,0 +1,149 @@
+
+    enterprise_customer = get_enterprise_customer(enterprise_customer_uuid)
+    return enterprise_customer is not None and any([
+        enterprise_customer.enforces_data_sharing_consent('at_login'),
```
I thought we were not planning to support 'at_login' for the new endpoint and are removing it later, @haikuginger @douglashall ?

UmanShahzad replied:
Personally, I didn't even know about this until going through @haikuginger's recent PR. I will be deleting any and all things related to `at_login` in this PR; thanks for noting :)

UmanShahzad commented:
I've removed `at_login` and `optional` consent logic / tests related to this app that I shouldn't have added in the first place... and added more test cases.

I just have to add the appropriate permission class (staff only), which I'll do tomorrow. For that, I'll either straight drop in the code that's written in my PR summary, or use the one from the edx-platform (but I feel safer doing the former, even if it leads to some repeat code, because of the potential to become an independent app).

UmanShahzad commented:
@haikuginger @brittneyexline I've added the permission classes, coverage for them, and changed the name of the `enterprise_customer` key to `enterprise_customer_uuid` to make it more clear what we want as input. I've also made the tests cleaner and more easily changeable in case some key's names change, etc..

brittneyexline commented on consent/helpers.py:
```diff
@@ -43,7 +43,7 @@ def consent_provided(user_id, course_id, enterprise_customer_uuid):
     provided = enterprise_course_enrollment is not None and enterprise_course_enrollment.consent_available
 
-    if enterprise_course_enrollment is None:
+    if enterprise_course_enrollment is None or not enterprise_course_enrollment.consent_available:
```
i think this should actually be `or enterprise_course_enrollment.consent_available is None`, since if it is false, that means they declined consent.

UmanShahzad replied:
Good point, done

UmanShahzad commented:
@brittneyexline @haikuginger Let's have a final look?

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,228 @@
+            self.REQUIRED_PARAM_USERNAME: username,
+            self.REQUIRED_PARAM_COURSE_ID: course_id,
+            self.REQUIRED_PARAM_ENTERPRISE_CUSTOMER: enterprise_customer_uuid,
+            self.CONSENT_EXISTS: response_status == 200,
```
Ah yep, did this in a rush, thanks!

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,228 @@
+        enterprise_course_enrollment = EnterpriseCourseEnrollment.objects.get_or_create(
+            enterprise_customer_user=enterprise_customer_user,
+            course_id=course_id
+        )[0]
```
:+1:

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,228 @@
+        if request.method == 'GET':
+            username = request.query_params.get(self.REQUIRED_PARAM_USERNAME, '')
+            course_id = request.query_params.get(self.REQUIRED_PARAM_COURSE_ID, '')
+            enterprise_customer_uuid = request.query_params.get(self.REQUIRED_PARAM_ENTERPRISE_CUSTOMER, None)
```
:+1:

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,228 @@
+    CONSENT_GRANTED = 'consent_provided'
+    CONSENT_REQUIRED = 'consent_required'
+
+    def get_required_query_params(self, request):
```
Yep, a 400-series error is returned if any of them are missing. I've added tests now to also cover the case that `course_id` is missing.

haikuginger replied:
Ohhh. I must have been getting an error-code response, but the overall response was still being rendered fully. Maybe we should return an actual structurally-different response if the parameters are missing, rather than proceeding to do the database queries with the empty values but just flagging with a 400 code.

UmanShahzad replied:
Sure, that's possible too. A body with just an error message?

UmanShahzad replied:
I've changed it to now return an error message on 400's. But a 404 (for GET when params are good but object doesn't exist) will still just return a normal body.

UmanShahzad commented on consent/helpers.py:
```diff
@@ -0,0 +1,146 @@
+    enterprise_course_enrollment = get_enterprise_course_enrollment(user_id, course_id, enterprise_customer)
+    provided = enterprise_course_enrollment is not None and enterprise_course_enrollment.consent_available
+
+    if enterprise_course_enrollment is None or enterprise_course_enrollment.consent_available is None:
```
:+1:

UmanShahzad commented on consent/api/v1/views.py:
```diff
@@ -0,0 +1,228 @@
+
+    """
+
+    permission_classes = [permissions.IsStaff, ]
```
Will give it a shot.

UmanShahzad commented:
Squashed on (`abfeae3`)[https://github.com/open-craft/edx-enterprise/commit/abfeae354c49d7a0b77e5261476c2c3712ae8b68].

UmanShahzad commented:
@haikuginger I hope this should have addressed your nits! See the commit in the above comment specifically for what addressed the review.

UmanShahzad commented:
Squashed on [1f85456](https://github.com/open-craft/edx-enterprise/commit/1f85456a38fc9283738bff335532546d5238a3eb).

UmanShahzad commented:
@haikuginger Ok, let me know if the error-case is good now.

UmanShahzad commented:
@brittneyexline Possible to get a final review from you? :)

haikuginger commented on docs/conf.py:
```diff
@@ -463,7 +463,7 @@
     'python': ('https://docs.python.org/3.5', None),
     'django': ('https://docs.djangoproject.com/en/1.10/', 'https://docs.djangoproject.com/en/1.10/_objects/'),
-    'model_utils': ('https://django-model-utils.readthedocs.io/en/latest/', None),
+    'model_utils': ('https://django-model-test_utils.readthedocs.io/en/latest/', None),
```
I think this is an accidental change.

UmanShahzad replied:
Ugh. Yes it is... the IDE did it. My fault for not carefully reading the diff, though.

UmanShahzad commented:
Squashed from 815fbb30ddbec2b5c4c47be1cdc0d654f0a500f4.
